File: Noxa.Emulation.Psp.Cpu.R4000DynRec/Tracer.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#define FLUSHONWRITE

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Diagnostics;

namespace Noxa.Emulation.Psp.Cpu
{
	static class Tracer
	{
		private static FileStream _stream;
		private static StreamWriter _writer;

		public static void OpenFile( string fileName )
		{
			_stream = File.OpenWrite( fileName );
			_writer = new StreamWriter( _stream );
		}

		public static void CloseFile()
		{
			if( _writer != null )
				_writer.Dispose();
			_writer = null;
			_stream = null;
		}

		public static void WriteLine( string line )
		{
			Debug.Assert( _writer != null );

			_writer.WriteLine( line );
#if FLUSHONWRITE
			Flush();
#endif
		}

		public static void WriteLine( string format, params object[] arg )
		{
			Debug.Assert( _writer != null );

			_writer.WriteLine( format, arg );
#if FLUSHONWRITE
			Flush();
#endif
		}

		public static void Flush()
		{
			Debug.Assert( _writer != null );
			_writer.Flush();
		}
	}
}

File: trunk/Shared/Source/Tracer.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#define WIN32_LEAN_AND_MEAN
#include <Windows.h>
#include <string>
#include "Tracer.h"

using namespace Noxa::Emulation::Psp;

HANDLE Tracer::_file;

void Tracer::OpenFile( const char* fileName )
{
	_file = ::CreateFileA( fileName, GENERIC_WRITE, FILE_SHARE_READ, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_TEMPORARY, NULL );
}

void Tracer::CloseFile()
{
	if( _file != NULL )
		::CloseHandle( _file );
	_file = NULL;
}

#pragma warning( disable: 4949 )

#pragma unmanaged
void Tracer::WriteLine( const char* line )
{
	if( _file == NULL )
		return;
	int length = strlen( line );
	int dummy;
	::WriteFile( _file, line, length, ( LPDWORD )&dummy, NULL );
}

void Tracer::WriteBytes( const byte* buffer, int length )
{
	if( _file == NULL )
		return;
	int dummy;
	::WriteFile( _file, buffer, length, ( LPDWORD )&dummy, NULL );
}

void Tracer::Flush()
{
	::FlushFileBuffers( _file );
}
#pragma managed
